### PR TITLE
sql: support <type> ARRAY syntax in CREATE

### PR DESCRIPTION
**Pull request by justinj**
This commit implements defining an array column type by, for example INT
ARRAY rather than INT[].

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parser/sql.y, line 3724 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455#-KqjxDNCd0d42-vnwnfe:-KqjxDNCd0d42-vnwnff:bs5fg7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/51bff2daba419a10b040ec87ef4df2cd11607f50/pkg/sql/parser/sql.y#L3724)):*
> ```Yacc
>   }
>   // SQL standard syntax, currently only one-dimensional
> | simple_typename ARRAY '[' ICONST ']' { return unimplementedWithIssue(sqllex, 2115) }
> ```

While we're here, let's change this to issue number #17156 to be consistent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455#-:-Kqjy7rifPciCsgK7qRF:bnfp4nl)*

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -3722,7 +3722,14 @@ typename:
   }
   // SQL standard syntax, currently only one-dimensional
 | simple_typename ARRAY '[' ICONST ']' { return unimplementedWithIssue(sqllex, 2115) }
```
Comment (knz) [inline pkg/sql/parser/sql.y]: What about the previous grammar rule then?

Comment (nvanbenschoten) [inline pkg/sql/parser/sql.y]: @knz is that a question to me about my previous comment? Array bounds are left unimplemented in the `opt_array_bounds` rule as well.

Comment (knz) [inline pkg/sql/parser/sql.y]: It seems to me that in the same way we support `VARCHAR(n)` and ignore `n`, we could support `ARRAY[n]` and ignore `n` too.

**Comment by justinj** [comment]
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/sql.y, line 3724 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455#-KqitZcHWAP3ZYlmJYWu-r1-3724:-KqwsfOP4xtqYsDmtL5V:b-ixarsi) ([raw file](https://github.com/cockroachdb/cockroach/blob/51bff2daba419a10b040ec87ef4df2cd11607f50/pkg/sql/parser/sql.y#L3724)):*
Is that true? I thought `VARCHAR(n)` actually did truncate/restrict its values:
```
> SELECT 'asdfasdfasdf'::VARCHAR(3);
+----------------------------+
| 'asdfasdfasdf'::VARCHAR(3) |
+----------------------------+
| asd                        |
+----------------------------+
(1 row)
> create table x (a varchar(3));
CREATE TABLE
> insert into x values ('hello');
pq: value too long for type STRING(3) (column "a")
```

That said, we could in theory ignore the value for arrays, since that's actually what Postgres does. I'm not convinced that's desirable, though. I think the conclusion that @jordanlewis and I came to was that this seemed like an odd behaviour of Postgres. My opinion isn't especially strong though, if you feel otherwise.

---

*[pkg/sql/parser/sql.y, line 3724 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455#-KqjxDNCd0d42-vnwnfe:-KqwuHRrAuFmYDPgLcLN:b-oz8tyz) ([raw file](https://github.com/cockroachdb/cockroach/blob/51bff2daba419a10b040ec87ef4df2cd11607f50/pkg/sql/parser/sql.y#L3724)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

While we're here, let's change this to issue number #17156 to be consistent.
</blockquote></details>

Good call!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455)*

**Comment by jordanlewis** [comment]
:lgtm:

---

Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455#-:-KqxqiKvli4TwNUXFtIJ:bnfp4nl)*

**Comment by justinj** [comment]
Ok, at @knz's suggestion we now allow the bounds but throw them away a la postgres. PTAL!

**Comment by knz** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17455#-:-KrG6G7ZtLVhNxJQYaIx:bnfp4nl)*

**Pull request merged**